nucleof072rb/Core/Src/main.c

```diff
@@ -1,181 +1,216 @@
-  /* USER CODE BEGIN 6 */
-  /* User can add his own implementation to report the file name and line number,
-     ex: printf("Wrong parameters value: file %s on line %d\r\n", file, line) */
-  /* USER CODE END 6 */
-}
-#endif /* USE_FULL_ASSERT */
-
-/************************ (C) COPYRIGHT STMicroelectronics *****END OF FILE****/
+/* USER CODE BEGIN Header */
+/**
+  ******************************************************************************
+  * @file           : main.c
+  * @brief          : Main program body
+  ******************************************************************************
+  * @attention
+  *
+  * <h2><center>&copy; Copyright (c) 2021 STMicroelectronics.
+  * All rights reserved.</center></h2>
+  *
+  * This software component is licensed by ST under BSD 3-Clause license,
+  * the "License"; You may not use this file except in compliance with the
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+//SPI_HandleTypeDef hspi1;
+//TIM_HandleTypeDef htim1;
+
+/* USER CODE BEGIN PV */
+uint8_t adc_tx_data[3] = {0x1, 0x10000000, 0x0};  // Reading from channel 0
+uint8_t adc_rx_data[3];     // Data received from ADC
+uint32_t adc_value = 0;      // 10-bit ADC value
+uint32_t pwm_value = 0;
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+void read_adc_spi(void) {
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);  // Set CS low
+
+    HAL_SPI_TransmitReceive(&hspi1, adc_tx_data, adc_rx_data, 1, HAL_MAX_DELAY);
+
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);  // Set CS high
+
+
+    adc_value = ((adc_rx_data[1] & 0x03) << 8) | adc_rx_data[2]; // get only 10 last bits (refer to datasheet
+}
+
+void adc_to_pwm() {
+
+    pwm_value = ((adc_value * (2000 - 1000)) / 1023) + 1000;

```

Comment:
the `1000` doesn't map to 1ms here since you have a clock speed of 48MHz/16(prescaler) = 3MHz. (1sec/3M)*1k != 1ms